addons/binding/org.openhab.binding.mihome/src/main/java/org/openhab/binding/mihome/internal/MiHomeHandlerFactory.java

```diff
@@ -0,0 +1,86 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.mihome.internal;
+
+import static org.openhab.binding.mihome.MiHomeBindingConstants.*;
+
+import java.util.HashMap;
+import java.util.Hashtable;
+import java.util.Map;
+import java.util.Set;
+
+import org.eclipse.smarthome.config.discovery.DiscoveryService;
+import org.eclipse.smarthome.core.thing.Bridge;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandlerFactory;
+import org.eclipse.smarthome.core.thing.binding.ThingHandler;
+import org.openhab.binding.mihome.handler.MiHomeGatewayHandler;
+import org.openhab.binding.mihome.handler.MiHomeSubdevicesHandler;
+import org.openhab.binding.mihome.internal.discovery.MiHomeSubdeviceDiscoveryService;
+import org.osgi.framework.ServiceRegistration;
+
+import com.google.common.collect.Sets;
+
+/**
+ * The {@link MiHomeHandlerFactory} is responsible for creating things and thing
+ * handlers.
+ *
+ * @author Svilen Valkanov - Initial contribution
+ */
+public class MiHomeHandlerFactory extends BaseThingHandlerFactory {
+
+    @SuppressWarnings("rawtypes")
```

Comment:
Because of compatibility issues.

SuppressWarnings is needed for BundleContext.getServiceReference() and BundleContext.registerService(), they don't have a typed version in the OSGi 4.2 specification and in some implementations as well.